### PR TITLE
Multiplayer browser - Display server latency

### PR DESCRIPTION
**Pull request by rob-v**
obrazok

UI is not blocked while pinging servers and each IP is pinged only once on each 'Refresh'.

+ added map title tooltip

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs [hunk 1]
```diff
@@ -403,7 +407,18 @@ void RefreshServerListInner(List<GameServer> games)
...
+						var host = game.Address.Split(':')[0];
 						var item = ScrollItemWidget.Setup(serverTemplate, () => currentServer == game, () => SelectServer(game), () => Join(game));
+
+						var block = item.GetOrNull("LATENCY");
+						if (block != null)
+						{
+							var pingTask = pingCache.GetOrAdd(host, key => new Lazy<Task<double>>(() => Task.Run(() => OpenRA.Server.Server.GetAveragePingTime(key, 4)))).Value;
+							block.IsVisible = () => pingTask.IsCompleted;
+							block.Get<ColorBlockWidget>("LATENCY_COLOR").GetColor = () => LobbyUtils.LatencyColor(
+								new Session.ClientPing() { Latency = (long)pingTask.Result });
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs]: This will create and discard roughly 600 ClientPings per second, which we obviously don't want to do.  This only needs to be evaluated once, when the task is complete.

**Comment by pchote** [comment]
You will also need to update the TD yaml.

**Comment by rob-v** [comment]
Updated.

**Comment by rob-v** [comment]
There is currently avg of 4 pings, but 2 could be sufficient and faster?
Wasn't it better to add new range/color to latency? (or make it stricter)

**Comment by pchote** [comment]
The colors represent "probably good" "probably okay most of the time" and "probably going to lag".  What other ranges are there to represent?

**Comment by rob-v** [comment]
I agree this is good for lobby. I realized, that IMO latency in server browser is different and more colors/seeing pings/stricter range would overall increase performance and experience. Players can choose from hundreds of servers (at least from empty to host a new game). There are several 40 ms and 280 ms servers, both green. Without a choice between them, the risk of caused lag lately is higher due to changing NW utilization (on system or outside) due to less reserve with 280ms server. This would have more impact in case server might adapt performance based on situation. 
Selecting better server can save also a lot of time and keep more players - when you join 280ms, it can happen you'll be kicked after 10 min waiting due to latency changing to yellow (and back). For players in a location where most servers have higher pings (currently no indication or green all) to find one good (with stable green in lobby) can be tiresome.

**Comment by kyrreso** [comment]
Should this in the future also show if it's a dedicated or not?

**Comment by rob-v** [comment]
I don't know, but this information is currently not available for client/game, web serve would need to be adapted also.

File: OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs [hunk 2]
```diff
@@ -444,7 +462,7 @@ void RefreshServerListInner(List<GameServer> games)
 						if (location != null)
 						{
 							var font = Game.Renderer.Fonts[location.Font];
-							var cachedServerLocation = GeoIP.LookupCountry(game.Address.Split(':')[0]);
```
Comment (chrisforbes) [inline OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs]: Was this supposed to be in a separate commit?

Comment (chrisforbes) [inline OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs]: Disregard.

**Comment by rob-v** [comment]
Please ignore it atm, I'm working on updated version which will ping only visible servers.

**Comment by rob-v** [comment]
Updated - only servers visible on screen are pinged and at most once (avg of 2 pings) in 20 s

**Comment by rob-v** [comment]
Not important, closing.

**Comment by pchote** [comment]
Adding to the future milestone, i'd like to revisit this in the future if we can find more time.

**Comment by teinarss** [comment]
After discussion on IRC. Using ICMP ping is probably not the proper way to do it.  So if no other solution is found this seams unlikely.
